src/ma/downscaler/downscale_supply_monthly_gen_to_hh.py

```diff
@@ -0,0 +1,367 @@
+        year = int(timestamp.year)
+        month = int(timestamp.month)
+
+        # For each tech
+        for tech in ProductionTechEnum:
+            tech_col = f"{tech}_mwh"
+
+            if tech_col not in row:
+                continue
+
+            grid_hh_value = row[tech_col]
+
+            # Skip if there's no generation for this tech/timestamp
+            if grid_hh_value <= 0:
+                continue
+
+            # Find the corresponding scaling factors
+            tech_scaling = scaling_df[
+                (scaling_df["year"] == year) & (scaling_df["month"] == month) & (scaling_df["tech"] == tech)
+            ]
+
+            # Apply scaling for each supplier
+            for _, scale_row in tech_scaling.iterrows():
+                supplier = scale_row["supplier"]
+                factor = scale_row["scaling_factor"]
+
+                # Calculate supplier's generation for this timestamp
+                supplier_hh_value = grid_hh_value * factor
+
+                # Add a record
+                records.append(
+                    {
+                        "timestamp": timestamp,
+                        "tech": tech,
+                        "supplier": supplier,
+                        "generation_mwh": supplier_hh_value,
+                    }
+                )
+
+    # Convert records to DataFrame
+    if not records:
+        return pd.DataFrame(columns=["timestamp", "tech", "supplier", "generation_mwh"])
+
+    result_df = pd.DataFrame(records)
+    return result_df
+
+
+def _validate_date_ranges(
+    start_datetime: pd.Timestamp,
+    end_datetime: pd.Timestamp,
+    grid_mix_hh: pd.DataFrame,
+    regos_data: pd.DataFrame,
+) -> bool:
+    """
+    Validates that the requested date range is available in both grid mix and REGOS data.
+
+    Parameters
+    ----------
+    start_datetime : pd.Timestamp
+        The start date for the analysis
+    end_datetime : pd.Timestamp
+        The end date for the analysis (exclusive)
+    grid_mix_hh : pd.DataFrame
+        The half-hourly grid mix data with DatetimeIndex
+    regos_data : pd.DataFrame
+        The REGOS certificate data with a 'month' column
+
+    Returns
+    -------
+    bool
+        True if all checks pass, otherwise raises ValueError with appropriate message
+    """
+    # Validate grid_mix_hh has DatetimeIndex
+    if not isinstance(grid_mix_hh.index, pd.DatetimeIndex):
+        raise ValueError("Grid mix data must have a DatetimeIndex")
+
+    error_messages = []
+
+    # Check grid mix data range
+    grid_min_date = grid_mix_hh.index.min()
+    grid_max_date = grid_mix_hh.index.max()
+
+    if start_datetime < grid_min_date:
+        error_messages.append(
+            f"Start date {start_datetime} is before available grid mix data (starting {grid_min_date})."
+        )
+    if end_datetime > grid_max_date:
+        error_messages.append(f"End date {end_datetime} is after available grid mix data (ending {grid_max_date}).")
```

Comment:
I have a minor personal preference for succinct exception checking - feel free to disregard if you have a strong opinion.



```suggestion
    if start_datetime < grid_mix_hh.index.min() or end_datetime > grid_mix_hh.index.max():
        error_messages.append("...")
```

Reply:
Much neater